src/modes/ModeEstimation.jl

```diff
@@ -440,3 +442,190 @@ function _optimize(
 
     return ModeResult(vmat, M, -M.minimum, f)
 end
+
+
+#################################################
+# Generic optimisation objective initialisation #
+#################################################
+
+function (f::Turing.OptimLogDensity)(G, z)
+  spl = DynamicPPL.SampleFromPrior()
+  
+  # Calculate log joint and the gradient
+  l, g = Turing.gradient_logp(
+      z, 
+      DynamicPPL.VarInfo(f.vi, spl, z), 
+      f.model, 
+      spl,
+      f.context
+  )
+
+  # Use the negative gradient because we are minimizing.
+  G[:] = -g
+
+  return nothing
+end
+
+
+
+function transform!(f::OptimLogDensity)
+  spl = DynamicPPL.SampleFromPrior()
+
+  ## Check link status of vi in OptimLogDensity
+  linked = DynamicPPL.islinked(f.vi, spl) 
+
+  ## transform into constrained or unconstrained space depending on current state of vi
+  if !linked
+    f.vi[spl] = f.vi[DynamicPPL.SampleFromPrior()]
+    DynamicPPL.link!(f.vi, spl)
+  else
+    DynamicPPL.invlink!(f.vi, spl)
+  end
+
+  return nothing
+end
+
+
+function transform2constrained(par::AbstractArray, vi::DynamicPPL.VarInfo)
```

Comment:
```suggestion
function transform2constrained(vi::DynamicPPL.VarInfo, par::AbstractArray)
```

Reply:
How come you want to change the order of arguments? I had put it this way so an inplace ! version which mutating par follows naturally

Reply:
Ah my bad, I somehow thought it is called `transform2constrained!` here as well (and would mutate `vi`). BTW I don't think the in-place versions are particularly useful since they do not operate in-place actually - they still create an intermediate new array (with the out-of-place version) which is then copied. The usual approach when you want to have both is to implement an in-place version without any intermediate allocations, and then create a new separate output array first for the out-of-place method before calling the in-place implementation.

A different criticism here is that one has to provide `vi` at all. I imagined something like
```julia
struct TransformFunction{V}
  vi::V
end

function (t::TransformFunction)(p::AbstractArray)
    vi = t.vi
	...
end 
```
Then you also don't have to define the new closures in the instantiate_.... functions.